netket/optimizer/rgn/rgn_pytree_logic.py

```diff
@@ -0,0 +1,358 @@
+    """
+    return jax.tree_map(partial(subtract_mean, axis=0), oks)  # MPI
+
+
+def tree_mean(oks: PyTree) -> PyTree:
+    """
+    subtract the mean with MPI along axis 0 of every leaf
+    """
+    return jax.tree_map(partial(mean, axis=0), oks)
+
+def jacobian_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array, chunk_size: int = None
+) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→R or holomorphic C→C, so single grad is enough
+
+    Args:
+        forward_fn: the log wavefunction ln Ψ
+        params : a pytree of parameters p
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+
+    def _jacobian_real_holo(forward_fn, params, samples):
+        y, vjp_fun = jax.vjp(single_sample(forward_fn), params, samples)
+        res, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+        return res
+
+    return vmap_chunked(
+        _jacobian_real_holo, in_axes=(None, None, 0), chunk_size=chunk_size
+    )(forward_fn, params, samples)
+
+
+def jacobian_cplx(
+    forward_fn: Callable,
+    params: PyTree,
+    samples: Array,
+    chunk_size: int = None,
+    _build_fn: Callable = partial(jax.tree_multimap, jax.lax.complex),
+) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→C, backpropagates 1 and -1j
+
+    Args:
+        forward_fn: the log wavefunction ln Ψ
+        params : a pytree of parameters p
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+
+    def _jacobian_cplx(forward_fn, params, samples, _build_fn):
+        y, vjp_fun = jax.vjp(single_sample(forward_fn), params, samples)
+        gr, _ = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+        gi, _ = vjp_fun(np.array(-1.0j, dtype=jnp.result_type(y)))
+        return _build_fn(gr, gi)
+
+    return vmap_chunked(
+        _jacobian_cplx, in_axes=(None, None, 0, None), chunk_size=chunk_size
+    )(forward_fn, params, samples, _build_fn)
+
+centered_jacobian_real_holo = compose(tree_subtract_mean, jacobian_real_holo)
+centered_jacobian_cplx = compose(tree_subtract_mean, jacobian_cplx)
+
+def rhessian_real_holo(
+    forward_fn: Callable, params: PyTree, samples: Array, connected_samples: Array, matrix_elements: Array, chunk_size: int = None
+) -> PyTree:
+    """Calculates Jacobian entries by vmapping grad.
+    Assumes the function is R→R or holomorphic C→C, so single grad is enough
+
+    Args:
+        forward_fn: the log wavefunction ln Ψ
+        params : a pytree of parameters p
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+
+    def _rhessian_real_holo(logpsi, params, σ, σp, mels):
+        y, vjp_fun = jax.vjp(logpsi, params, σp)
+        scale = mels*jnp.exp(logpsi(params,σp) - logpsi(params,σ))
```

Comment:
```suggestion
        scale = mels*jnp.exp(y - logpsi(params,σ))
```
This is exactly what I meant.

Reply:
It just needs to be integrated with `expect_and_grad` and so probably moved to the driver